### ISSUE TITLE
expo-image & flashlist: component recycling

### ISSUE THREAD

**Issue by NTag** [expo/expo#21211]
### Minimal reproducible example

Using expo-image inside a flashlist

### Summary

[@shopify/flash-list](https://shopify.github.io/flash-list/docs/) is re-using the JS components. It means that if you're using expo-image inside the list, the same component will be re-used and just its source will change.

However, expo-image displays the previous source until the new one has been fully loaded. I know it's a feature but in this use-case it's confusing for the user as after scrolling fast they end up seeing old images whereas it should be better to display nothing.

How would you advise handling this use-case? One can add a `key` prop on the `expo-image` component but it goes against flash-list rules and optimizations. Do you see other options? What would be your recommandation?

### Environment

Expo SDK 47, expo-image RC.0, flashlist 1.4.0

**Comment by terrysahaidak** [comment]
@NTag did you try to provide `cacheKey` as part of source object?

**Comment by NTag** [comment]
I don't think it would work:
<img width="497" alt="CleanShot 2023-02-15 at 12 30 30@2x" src="https://user-images.githubusercontent.com/906276/219015769-051fdaed-85bc-4ccc-8b4a-3f5486230c76.png">

As the default is the URL, if the URL changes, the cacheKey changes, so I don't see what I could do better

**Comment by BubbleTrouble14** [comment]
Expieriencing the same issue. As I am using flashlist with the new expo-image as react-native-fast-image is no longer mantained. Any workaround or fix available ? https://gyazo.com/4475931ac2402edd442d87785e3b4feb

**Comment by tsapeta** [comment]
I'll take a look at this issue before the stable 1.0.0 release.
Could you please add a minimal reproducible example next time so our bot doesn't close the issue automatically? That really helps us solve and deliver fixes faster.

**Comment by BubbleTrouble14** [comment]
Yes, already got a minimal repo. Will share it tommorow 👍

**Comment by BubbleTrouble14** [comment]
@tsapeta https://github.com/BubbleTrouble14/ExpoImageIssue Hopefully this helps.

**Comment by hirbod** [comment]
FYI: Adding the key basically breaks recycling, so that should not be done, even if it might fix the issue in the first place.
I'd like to quote one of the main devs of FlashList here, @naqvitalha: (copied from Discord)

> Usually, libraries should reset their view when URLs change. A similar problem can be seen with Image on iOS. It keeps displaying the old image until the new one is downloaded. Ideally, it should instantly switch to a placeholder and then load.

Since expo-image is recycled (and thus most likely already rendered), the native side should detect the change in either cache key or URL and instantly switch to a placeholder, defaultSource, or just render nothing. It might be worth considering this as opt-in, but there should be such a function to support recycling more properly.

**Comment by terrysahaidak** [comment]
I saw this same behavior with RN's Image in my project.
Also, I saw it many times on TikTok's direct messages list on Android. So the problem could be more fundamental than just a bug in the library.
Also, I just remembered now I've seen this same bug at Rainbow's asset list (screen with all your wallet assets) which is using recyclerlistview and FastImage.

**Comment by hirbod** [comment]
@terrysahaidak, that's not a bug per se; it is how the native side treats the element. The Image is already rendered and its prop has changed. Currently, it keeps displaying the old image while the new one is being fetched. Once the download is finished, it will change the source. The fix should be pretty straightforward, allowing for an instant change when a change in the props is detected by opting in.

**Comment by terrysahaidak** [comment]
But can we somehow force expo-image to render placeholder instead?

**Comment by hirbod** [comment]
Currently, it's not possible to achieve this functionality, which is why I proposed an opt-in solution. This would force the native side to render the view twice; once when there's a change in the URL or cache key, and again when the final image is fetched.

> Since in my case it happens when I add a new item to the beginning of the list. It reuses the same view from the end of the list, so draw distance won't help.

This approach would fix the above issue. FlashList currently recycles one of the last views, which may display an outdated image until the new one is fetched. The native side should detect this change and treat it as a new image, unloading the old one and replacing it with the new one instantly once fetched. The current behavior is appropriate for most use cases, such as when updating a profile picture where it's appropriate to show a spinner and then update the source once the image has been fetched. However, this "keep image until new fetched" technique is unsuitable for recycling.

**Comment by tsapeta** [comment]
We're going to introduce another prop called `recyclingKey` (still not sure if that's a good name though) which, when changed, will reset the image view content instead of showing the previous source. You could set it to a string that contains the item id or the row + column indexes or just use the source uri if you don't need the source-to-source transitioning behavior besides scrolling.

In other words, it's going to work like `key` but it doesn't remount the image view itself but only its content. I believe it should be a bit faster since the view can be recycled.

Does this solution sound good to you all? If so, we might release a new `rc` version today and once you confirm it solved the problem we'll make it stable 😄

**Comment by NTag** [comment]
That sounds very good to me! Thank you very much 🙏

**Comment by tsapeta** [comment]
`expo-image@1.0.0-rc.2` is out 🙂 Looking forward to your feedback 😊

**Comment by NTag** [comment]
Actually I wonder if it's not when I use the placeholder prop, maybe your PR needed something like `.placeholderImage = nil` too to reset it too?

**Comment by BubbleTrouble14** [comment]
Working on my end with the example I shared 👍

**Comment by hirbod** [comment]
I tested it and it works fine here. Just make sure that `recyclingKey` is truly unique and that you've updated the pods correctly.

However, in DEV MODE, I noticed that when the JS FPS drops, FlashList may not be fast enough to pass new props to the recycled image, which could briefly lead to this issue. But this is not an issue with expo-image. I couldn't reproduce this in release/prod. So I'd say that this issue is properly fixed.

**Comment by NTag** [comment]
You both tested also the placeholder prop and it works? Damn, I'll double-check then

**Comment by BubbleTrouble14** [comment]
I can give it a try.

**Comment by tsapeta** [comment]
@NTag, I think reseting the placeholder is not necessary, we store it as a property for some reasons but it doesn't work like `view.image = image` – the setter doesn't change the view's content.
I'd like to try it out with the placeholder anyway, do you maybe have a minimal reproducible example? Are your placeholders different for each image or you use just the one for all?

**Comment by fobos531** [comment]
@tsapeta Is it safe to pass a `recyclingKey` (and set it to the image URI) even though we may not use the component in a recyclable list? Also, would transitions from blurhash to the image work fine if we use the image source/URI as the `recyclingKey`?

**Comment by hirbod** [comment]
@fobos531 there is no need to pass recyclingKey outside of a recycler. But it will also not do you any harm.

**Comment by glundgrenm** [comment]
I've tried here and reclyclingKey does not work if we use an image inside a renderItem on FlashList.

**Comment by hirbod** [comment]
Did you pass `recyclingKey` to `Image` or did you try to pass it on a wrong component? Forget about regular `key`, this breaks recycling. I can't confirm though, recyclingKey is working fine here in our renderItem.

**Comment by glundgrenm** [comment]
I've added to an Image (had to rollback from expo-image, it was breaking in android) in an custom Avatar component inside the renderItem component. I think it didn't broke recycling in Flashlist, the performance seems ok. How can i test if its recycling?

**Comment by hirbod** [comment]
Adding `key` to image breaks recycling for that component, as it will be recreated every time instead of being reused. Thats how `key` works and why it should be avoided. It might not be a huge deal though but perf should be better with recyclingKey and expo-image instead.

**Comment by kayode0x** [comment]
> But can we somehow force expo-image to render placeholder instead?
> 
> Since in my case it happens when I add a new item to the beginning of the list. It reuses the same view from the end of the list, so draw distance won't help.

Hey @terrysahaidak, I ran into the same issue, were you able to fix this?

**Comment by kayode0x** [comment]
I found a fix for this, sometimes all you probably need is a long nap and some documentation.

TLDR; I'm trying to add items to the top of a list, and on creation the new item would take the image from the last item on the list, not only buggy, but also annoying!

Following @hirbod's response 

> it may be possible to solve the issue using prepareForLayoutAnimationRender()

I couldn't find a way to test this because the item creation logic was in a context far away from the Flashlist, but re-reading the docs, I found a section on [`recycling.`](https://shopify.github.io/flash-list/docs/recycling)

As I understand it, the concept was to compare the current items to the previously rendered items, and if there was a change, the items needed to be updated.

Here's what fixed the problem on my end:

```
function RenderItems({ items, ListHeader, ListEmpty }: IRenderItems) {
    const listRef = useRef<FlashList<Item[]>>(null);
    const lastDataRef = useRef(items);
    const [data, setData] = useState(items);

    if (lastDataRef.current !== items) {
        // treats the new item as new ones, don't recycle.
        listRef.current?.prepareForLayoutAnimationRender();

        // items changed, so update the list.
        setData(items);
        lastItemsRef.current = items;
    }

    const keyExtractor = useCallback((item: IBookmark) => {
        return item.id;
    }, []);
 
    const renderItem = useCallback(({ item }: IRenderItem) => (
        <View>
            <Image
                  contentFit="cover"
                  source={item.image}
                  recyclingKey={item.id}
            />
            
            <Text>Item: {item.name}</Text>
        </View>
    ), []);

    return (
        <FlashList
            ref={ref}
            data={data}
            renderItem={renderItem}
            {...}
        />
    )
}
```

**Comment by tsapeta** [comment]
I guess the original issue has been resolved with the `recyclingKey` prop, so I'm closing it 🙂

**Comment by JacobJaffe** [comment]
Is there a minimal example of `recyclingKey` working with `placeholder`? I'm seeing blurhash placeholders not resizing correctly on recycled views

**Comment by tsapeta** [comment]
@JacobJaffe I suspect that might be an issue. Do you maybe have a minimal reproducible example so that we can investigate this faster? If so, could you please open a new issue?

**Comment by JacobJaffe** [comment]
Hi @tsapeta 

Here's a minimal reproduction:
https://github.com/JacobJaffe/reproduction-expo-image-blurhash-recycler

This is verifiable on both physical device and simulator running Expo Go.

I've included a video in the README of the repo of the issue occurring -- however, it's worth noting this is slightly different behavior than what I'm seeing in my app:
In the video of the reproduction, some blurhashes aren't even rendering at all, and then appear later when recycled as small versions. I haven't seen that happen in my use case. I've only seen the blurhashes becoming too small, but always appearing correctly in the initial, non recycled views.

**Comment by RRaideRR** [comment]
Also seeing the problem that @JacobJaffe describes 😇😬

**Comment by lewisd1996** [comment]
The issue i am facing is when using placeholder images, items in the flash list seem to change and duplicate, showing duplicate and incorrect images in other list items.

I have no `key` prop on the rendered items, I've tried using `recycleKey` but it has no affect. Removing the placeholders fixes the issue, but ideally would like to show blurred thumbnails for better UX.

This all seems fine on iOS for me, but on Android is a problem.

**Comment by JacobJaffe** [comment]
> The issue i am facing is when using placeholder images, items in the flash list seem to change and duplicate, showing duplicate and incorrect images in other list items.
> 
> I have no `key` prop on the rendered items, I've tried using `recycleKey` but it has no affect. Removing the placeholders fixes the issue, but ideally would like to show blurred thumbnails for better UX.
> 
> This all seems fine on iOS for me, but on Android is a problem.

@lewisd1996 Can you share what your `renderItem` looks like?

**Comment by lewisd1996** [comment]
> > The issue i am facing is when using placeholder images, items in the flash list seem to change and duplicate, showing duplicate and incorrect images in other list items.
> > I have no `key` prop on the rendered items, I've tried using `recycleKey` but it has no affect. Removing the placeholders fixes the issue, but ideally would like to show blurred thumbnails for better UX.
> > This all seems fine on iOS for me, but on Android is a problem.
> 
> @lewisd1996 Can you share what your `renderItem` looks like?

Sure thing, here is my component:

```
<StyledView className="h-full w-full px-4 pt-6">
  {!itemsInitialised ? (
    <FlashList
      showsVerticalScrollIndicator={false}
      className="h-full w-full"
      numColumns={2}
      ItemSeparatorComponent={() => <StyledView className="h-4" />}
      estimatedItemSize={370}
      renderItem={() => (
        <StyledView className="items-center justify-center px-2">
          <RectangleSkeleton width={width / 2 - 28} height={370} />
        </StyledView>
      )}
      data={new Array(8).fill({ id: 1 })}
    />
  ) : (
    <FlashList
      className="h-full w-full"
      data={setItems}
      estimatedItemSize={370}
      ItemSeparatorComponent={() => <StyledView className="h-4" />}
      numColumns={2}
      showsVerticalScrollIndicator={false}
      onEndReached={() => {
        if (hasMore && !isFetching && setItems && setItems.length > 0) {
          getMoreSetItems(setItems.length);
        }
      }}
      onEndReachedThreshold={0.2}
      renderItem={({ item }: { item: IFeedItem }) => (
        <FeedListItem item={item} />
      )}
    />
  )}
</StyledView>
```

And render item renders `FeedListItem`: 

```
function FeedListItem({ item }: { item: IFeedItem }) {
  const router = useRouter();

  return (
    <TouchableOpacity
      onPress={() => {
        router.push(`/post/${item?.id}`);
      }}
      className="mx-2 flex-1"
    >
      <Image
        contentFit="contain"
        placeholder={item?.images?.small}
        placeholderContentFit="cover"
        recyclingKey={item?.id}
        source={item?.images?.large}
        style={{
          aspectRatio: 3 / 4.2,
        }}
      />
      <StyledText
        numberOfLines={1}
        className="font-poppins-semibold mt-2 text-lg"
      >
        {item?.name}
      </StyledText>
    </TouchableOpacity>
  );
}
```

**Comment by abayomi185** [comment]
Hi, are there any updates or workarounds on the wrongly sized blurhash placeholder after the components have been recycled?

**Comment by JacobJaffe** [comment]
> Hi, are there any updates or workarounds on the wrongly sized blurhash placeholder after the components have been recycled?

@abayomi185 please see https://github.com/expo/expo/issues/22206 for the workarounds here, specifically: https://github.com/expo/expo/issues/22206#issuecomment-1551279377

**Comment by abayomi185** [comment]
Thanks! @JacobJaffe
